packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts

```diff
@@ -22,45 +23,156 @@ export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider
-    return super.update(data, context);
+    const objectId = new ObjectId(idField.value);
+
+    const startTime = Date.now();
+
+    const RETRY_LIMIT = 3000; // 3 seconds
+
+    while ((Date.now() - startTime) < RETRY_LIMIT) {   // Commented for debugging
+    // while (Date.now()) {
+
+      this.fieldTransformMap[TransformType.UPDATE]
+        .forEach((f: FieldTransform) => {
+          data[f.fieldName] = f.transform(data[f.fieldName]);
+        });
+
+      const updateFilter: any = { _id: objectId, _deleted: { $ne: true } };
+      const updateSets = Object.assign({}, data);
+      // eslint-disable-next-line @typescript-eslint/tslint/config
+      delete updateSets[this.tableMap.idField];
+
+      if (this.conflictEngine !== undefined) {
+        
+        const { conflictFieldName, next } = this.conflictEngine;
+        const currentState = updateSets[conflictFieldName];
+        updateFilter[conflictFieldName] = currentState;
+
+        // eslint-disable-next-line @typescript-eslint/tslint/config
+        delete updateSets[conflictFieldName]
+
+        updateSets[conflictFieldName] = next(currentState, updateSets);
+      }
+
+      const updateResult = await this.db.collection(this.collectionName).findOneAndUpdate(updateFilter, {
+        $set: updateSets
+      },{
+        w: "majority",
+        returnOriginal: false
+      })
+
+      if (!updateResult.value) {
+        // There is conflict or document doesn't exist
+  
+        // Check if document exists
+        const serverData = await this.db.collection(this.collectionName).findOne({ _id: objectId, _deleted: { $ne: true } })
+        if (!serverData){
+          throw new NoDataError(`Cannot update ${this.collectionName}`);
+        }
+
+        const { conflictFieldName, resolveConflicts } = this.conflictEngine;
+
+        let conflictMap:ConflictStateMap = {
+          serverState: serverData,
+          clientState: Object.assign({}, updateSets, { _id: objectId, [conflictFieldName]: updateFilter[conflictFieldName]})
+        }
+  
+        // Resolve conflicts/Reject Update:
+        if (resolveConflicts !== undefined) {
+          conflictMap = resolveConflicts(Object.assign({}, serverData), data)
+        }
+
+        if (conflictMap !== undefined){
+          const serverState = this.mapFields(serverData)
+
+          throw new ConflictError({ serverState, clientState: data})
+        }
+
+
+        // No conflicts, bump conflictField
+        data[conflictFieldName] = serverData[conflictFieldName];
+
+        continue
+      }
+
+      return this.mapFields(updateResult.value)
+
+    }
+    
+    throw new NoDataError(`Cannot update ${this.collectionName}`);
   }
 
   public async delete(data: any, context: GraphbackContext): Promise<Type> {
-    const conflict = await this.checkForConflicts(data, context);
-    if (conflict !== undefined) {
-      throw new ConflictError(conflict);
-    }
+    // throw new Error('not implemented');
```

Comment:
That is the big stuff - as you might notice we do not support deletes in offix as most of the business use cases we had simply did not had actual delete operation so this problem never occurred, but as generic solution goes on we want to support it. Let's chat how this could work in meta language first.